rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,597 @@
+		evnt->res_50.native = PS3AV_RESBIT_1280x720P;
+		evnt->res_50.res_bits = 0xf;
+		evnt->res_vesa.res_bits = 1;
+
+		evnt->cs.rgb = 1;
+		evnt->cs.yuv444 = 1;
+		evnt->cs.yuv422 = 1;
+
+		evnt->color.blue_x = 0xFFFF;
+		evnt->color.blue_y = 0xFFFF;
+		evnt->color.green_x = 0xFFFF;
+		evnt->color.green_y = 0xFFFF;
+		evnt->color.red_x = 0xFFFF;
+		evnt->color.red_y = 0xFFFF;
+		evnt->color.white_x = 0xFFFF;
+		evnt->color.white_x = 0xFFFF;
+		evnt->color.gamma = 100;
+
+		evnt->supported_ai = 0;
+		evnt->speaker_info = 0;
+		evnt->num_of_audio_block = 0;
+
+		vuart.write_resp(pkt->hdr.cid, PS3AV_STATUS_SUCCESS, evnt.get(), sizeof(ps3av_get_monitor_info_reply));
+		return true;
+	}
+};
+
+struct av_get_hw_conf_cmd : public ps3av_cmd
+{
+	u16 get_size(vuart_av_thread &vuart, const void *pkt_buf) override
+	{
+		static_cast<void>(vuart);
+		static_cast<void>(pkt_buf);
+		return 8;
+	}
+
+	bool execute(vuart_av_thread &vuart, const void *pkt_buf, u32 reply_max_size) override
+	{
+		auto pkt = static_cast<const ps3av_header *>(pkt_buf);
+
+		if (reply_max_size < sizeof(ps3av_get_hw_info_reply) + sizeof(ps3av_pkt_reply_hdr))
+		{
+			vuart.write_resp(pkt->cid, PS3AV_STATUS_BUFFER_OVERFLOW);
+			return false;
+		}
+
+		// TODO:
+		auto out = std::make_unique<ps3av_get_hw_info_reply>();
+		out->num_of_hdmi = 1;
+		out->num_of_avmulti = 0;
+		out->num_of_spdif = 1;
+		out->resv = 0;
+
+		vuart.write_resp(pkt->cid, PS3AV_STATUS_SUCCESS, out.get(), sizeof(ps3av_get_hw_info_reply));
+		return true;
+	}
+};
+
+struct generic_reply_cmd : public ps3av_cmd
+{
+	u16 get_size(vuart_av_thread &vuart, const void *pkt_buf) override
+	{
+		static_cast<void>(vuart);
+		static_cast<void>(pkt_buf);
+		return 0;
+	}
+
+	bool execute(vuart_av_thread &vuart, const void *pkt_buf, u32 reply_max_size) override
+	{
+		auto pkt = static_cast<const ps3av_header *>(pkt_buf);
+
+		vuart.write_resp(pkt->cid, PS3AV_STATUS_SUCCESS);
+		return true;
+	}
+};
+
+error_code sys_uart_initialize(ppu_thread &ppu)
+{
+	ppu.state += cpu_flag::wait;
+
+	sys_uart.trace("sys_uart_initialize()");
+
+	static shared_mutex m{};
+	std::lock_guard lock(m);
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (vuart_thread.init)
+	{
+		return CELL_EPERM;
+	}
+
+	vuart_thread.init = 1;
 
 	return CELL_OK;
 }
 
-error_code sys_uart_receive(vm::ptr<void> buffer, u64 size, u32 unk)
+error_code sys_uart_receive(ppu_thread &ppu, vm::ptr<void> buffer, u64 size, u32 mode)
 {
-	sys_uart.todo("sys_uart_receive(buffer=*0x%x, size=0x%llx, unk=0x%x)", buffer, size, unk);
+	sys_uart.trace("sys_uart_receive(buffer=*0x%x, size=0x%llx, mode=0x%x)", buffer, size, mode);
 
-	return CELL_OK;
+	if (!size)
+	{
+		return CELL_OK;
+	}
+
+	if ((mode & ~1UL) != 0)
+	{
+		return CELL_EINVAL;
+	}
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (mode > 1)
```

Comment:
Isn't this check redundant? See https://github.com/RPCS3/rpcs3/pull/10515/files#diff-db0f3c8901d150fda343fe5993b4eb6df2c0de2ed1d595de589af2fa4a3f9c74R145
In truth I much prefer this check over the other one for fast reading.

Reply:
It is surprisingly